### ISSUE TITLE
Nomad client v1.5.6 doesn't join to the Nomad server cluster after enable ACL

### ISSUE THREAD

**Issue by LudwinMarquina20** [hashicorp/nomad#17437]
Hello everyone.
After enabling ACL on my cluster, the Nomad Clients are unable to reconnect to the server cluster. My Consul cluster have 1.15.1.
The implementation was done directly in Nomad v1.5.6 (Server and client).
The error I have is:
```
2023-06-06T13:36:07.414Z [WARN]  client.server_mgr: no servers available
    2023-06-06T13:36:07.469Z [WARN]  client.server_mgr: no servers available
    2023-06-06T13:36:07.487Z [WARN]  client.server_mgr: no servers available
    2023-06-06T13:36:07.540Z [WARN]  client.server_mgr: no servers available
    2023-06-06T13:36:07.540Z [ERROR] client.rpc: error performing RPC to server, deadline exceeded, cannot retry: error="no servers" rpc=Node.Register
    2023-06-06T13:36:07.540Z [DEBUG] client: registration waiting on servers
    2023-06-06T13:36:07.541Z [ERROR] client: error discovering nomad servers: error="client.consul: unable to query Consul datacenters: Get \"https://127.0.0.1:8501/v1/catalog/datacenters\": dial tcp 127.0.0.1:8501: connect: connection refused"
    2023-06-06T13:36:07.684Z [WARN]  client.server_mgr: no servers available
    2023-06-06T13:36:07.765Z [WARN]  client.server_mgr: no servers available
    2023-06-06T13:36:07.765Z [ERROR] client.rpc: error performing RPC to server, deadline exceeded, cannot retry: error="no servers" rpc=Node.GetClientAllocs
    2023-06-06T13:36:07.822Z [WARN]  client.server_mgr: no servers available
    2023-06-06T13:36:07.943Z [WARN]  client.server_mgr: no servers available
    2023-06-06T13:36:07.943Z [ERROR] client.rpc: error performing RPC to server, deadline exceeded, cannot retry: error="no servers" rpc=Node.UpdateAlloc
    2023-06-06T13:36:07.943Z [ERROR] client: error updating allocations: error="no servers"
```
My configuration files are the next:

- NOMAD - SERVER (nomad.hcl).

```
datacenter = "dc1"
data_dir = "/opt/nomad"
server {
  enabled = true
  bootstrap_expect = 3
  raft_protocol = 3
}
advertise {
  http = "{{GetInterfaceIP \"eth0\"}}"
  rpc  = "{{GetInterfaceIP \"eth0\"}}"
  serf = "{{GetInterfaceIP \"eth0\"}}"
}
tls {
  http = true
  rpc  = true
  ca_file   = "/etc/nomad.d/ca.pem"
  cert_file = "/etc/nomad.d/erver.pem"
  key_file  = "/etc/nomad.d/server-key.pem"
  verify_server_hostname = false
  verify_https_client    = false
}
acl {
  enabled = true
}
```

- CONSUL - SERVER (consul.hcl).
```

"node_name"= "hashi-server-dev-1",
  "server"= true,
  "data_dir"= "/opt/consul",
  "bind_addr"= "{{ GetInterfaceIP \"eth0\" }}",
  "rejoin_after_leave"= true,
  "client_addr"= "0.0.0.0",
  "bootstrap_expect"= 3,
  "retry_join" = ["provider=aws tag_key=auto-join tag_value=consul-dev"],
  "encrypt" = "pc08zoU1C+k=",
  "enable_local_script_checks"= true,
  "ui_config"= {
    enabled= true
    metrics_provider= "prometheus"
  },
  "performance"= {
    "raft_multiplier"= 1
  },
  "ports"= {
    "https"= 8501
  },
  "tls"= {
    defaults {
      "verify_incoming"= true,
      "verify_outgoing"= true,
      "ca_file"= "/etc/consul.d/ca.pem",
      "cert_file"= "/etc/consul.d/server.pem",
      "key_file"= "/etc/consul.d/server-key.pem"
    }
    internal_rpc {
      verify_server_hostname = true
    }
  }
  acl {
    enabled        = true
    default_policy = "deny"
    down_policy    = "extend-cache"
    tokens {
      default = "2d87f5af-c1a4-7502-2e60-6602f15886f6"
    }
  }
```

- NOMAD - CLIENT (nomad.hcl).

```
datacenter = "dc1"
data_dir = "/opt/nomad"
log_level = "DEBUG"
client {
  enabled = true
  options   = {
    "docker.auth.config"     = "/root/.docker/config.json"
    "docker.auth.helper"     = "ecr-login"
  }
  host_network "localhost" {
    interface = "lo"
  }
  host_volume "monolith_public_key" {
    path = "/data/monolith/files/oauth-public.key"
    read_only = false
  }
  host_volume "monolith_private_key" {
    path = "/data/monolith/files/oauth-private.key"
    read_only = false
  }
  host_volume "datadog_sock" {
    path = "/var/run/docker.sock"
    read_only = false
  }
  host_volume "datadog_proc" {
    path = "/proc/"
    read_only = false
  }
  host_volume "datadog_cgroup" {
    path = "/sys/fs/cgroup/"
    read_only = false
  }
}
acl {
   enabled = true
}
consul {
  address             = "127.0.0.1:8501"
  ssl                 = true
  ca_file   = "/etc/nomad.d/rca.pem"
  cert_file = "/etc/nomad.d/client.pem"
  key_file  = "/etc/nomad.d/client-key.pem"
  server_service_name = "nomad"
  client_service_name = "nomad-client"
  auto_advertise      = true
  server_auto_join    = true
  client_auto_join    = true
}

tls {
  http = true
  rpc  = true
  ca_file   = "/etc/nomad.d/referwell-hashi-cluster-ca.pem"
  cert_file = "/etc/nomad.d/referwell-hashi-cluster-client.pem"
  key_file  = "/etc/nomad.d/referwell-hashi-cluster-client-key.pem"
  verify_server_hostname = false
  verify_https_client    = false
}
telemetry {
  collection_interval = "1s"
  disable_hostname = true
  prometheus_metrics = true
  publish_allocation_metrics = true
  publish_node_metrics = true
}
```

- CONSUL - CLIENT (nomad.hcl).

```
{
    "encrypt" : "pc08zoU1C+k=",
    "bind_addr": "{{ GetInterfaceIP \"eth0\" }}",
    "node_name": "hashi-client-dev-1",
    "data_dir": "/opt/consul",
    "rejoin_after_leave": true,
    "enable_local_script_checks": true,
    "retry_join" : ["provider=aws tag_key=auto-join tag_value=consul-dev"],
    "verify_incoming": true,
    "verify_outgoing": true,
    "verify_server_hostname": true,
    "ca_file": "/etc/consul.d/ca.pem",
    "cert_file": "/etc/consul.d/client.pem",
    "key_file": "/etc/consul.d/client-key.pem",

    "acl": {
        "enabled": true,
        "default_policy": "deny",
        "down_policy": "extend-cache",
        "tokens": {
            "default": "d562fe45-faf5-dafd-7bf2-3c0368e9afc9"
        }
    }
}
```
I have created the following policies and tokens in my Consul cluster.

```
# consul acl token list
AccessorID:       6eb65d08-b1b5-20b5-dec4-2b0764c24bbf
SecretID:         d562fe45-faf5-dafd-7bf2-3c0368e9afc9
Description:      consul-client
Local:            false
Create Time:      2023-06-05 16:36:38.795775571 +0000 UTC
Policies:
   74b6958c-bd28-0250-ecfe-b3f91b77d380 - consul-client

[ec2-user@ip-10-102-2-234 ~]$ consul acl policy list
consul-client:
   ID:           74b6958c-bd28-0250-ecfe-b3f91b77d380
   Description:
   Datacenters:
dns-requests:
   ID:           56a6896e-069c-53ad-a901-740b38d10309
   Description:
   Datacenters:
global-management:
   ID:           00000000-0000-0000-0000-000000000001
   Description:  Builtin Policy that grants unlimited access
   Datacenters:

[ec2-user@ip-10-102-2-234 ~]$ consul acl policy read -id 74b6958c-bd28-0250-ecfe-b3f91b77d380
ID:           74b6958c-bd28-0250-ecfe-b3f91b77d380
Name:         consul-client
Description:
Datacenters:
Rules:
agent_prefix "hashi-client" {
  policy = "write"
}

node_prefix "hashi-client" {
  policy = "write"
}

key_prefix "app-configs" {
  policy = "read"
}

key_prefix "traefik" {
  policy = "read"
}

service_prefix "" {
  policy = "write"
}

```
Please support me in understanding what is happening

- Nomad version: 1.5.6
- Consul version: 1.15.1

Thanks.

**Comment by jrasell** [comment]
Hi @LudwinMarquina20 and thanks for posting this issue.

The log line includes an error `client.consul: unable to query Consul datacenters: Get \"https://127.0.0.1:8501/v1/catalog/datacenters\": dial tcp 127.0.0.1:8501: connect: connection refused` which indicates the Nomad client is unable to connect to the Consul agent locally on `127.0.0.1:8501` because the API is not available on that host and port. Could you confirm that the local Consul agent is running on `127.0.0.1:8501` where the Nomad client is running?

**Comment by jrasell** [comment]
@LudwinMarquina20 I am going to close this issue as we have not received a response. If you are still experiencing problems, please feel free to reopen this issue and supply the additional context previously requested.

**Comment by koslib** [comment]
Hi there - I'm having similar issues as the OP. I'm using just Nomad though, with no Consul involved at all. Btw, nomad `v1.6.3`.

My server has a minimal setup like this (`server.hcl`):

```
data_dir  = "/opt/nomad/data"
bind_addr = "0.0.0.0"

server {
  enabled          = true
  bootstrap_expect = 1
}

acl {
  enabled = true
}
```

and the `client.hcl` is also quite simple like this:

```
data_dir  = "/opt/nomad/data"
bind_addr = "0.0.0.0"
datacenter = "dc1"

client {
  enabled = true
  options {
    "driver.raw_exec.enable"    = "1"
    "docker.privileged.enabled" = "true"
  }
}

acl {
  enabled = true
}

consul {}
```

As you can see, I have explicitly set the consul block to an empty block, but apparently this doesn't get Consul disabled. And I end up with these logs:

```
Nov 14 09:46:01 quilkin-nomad-client-0 nomad[12245]:     2023-11-14T09:46:01.923Z [WARN]  client.server_mgr: no servers available
Nov 14 09:46:01 quilkin-nomad-client-0 nomad[12245]:     2023-11-14T09:46:01.924Z [ERROR] client.rpc: error performing RPC to server, deadline exceeded, cannot retry: error="no servers" rpc=Node.Register
Nov 14 09:46:01 quilkin-nomad-client-0 nomad[12245]:     2023-11-14T09:46:01.924Z [DEBUG] client: registration waiting on servers
Nov 14 09:46:01 quilkin-nomad-client-0 nomad[12245]:     2023-11-14T09:46:01.925Z [ERROR] client: error discovering nomad servers: error="client.consul: unable to query Consul datacenters: Get \"http://127.0.0.1:8500/v1/catalog/datacenters\": dial tcp 127.0.0.1:8500: connect: connection refused"
```

1. Why are calls to Consul made when I've attempted to disable it? That block does not support `enabled = false` either.
2. I am not sure I understand what RPC call failed, neither I can find any logs for it (even though I've set the log level of the client to `DEBUG`).

Posting it here in case more folks have similar issues and in hopes I can get some support on this!

**Comment by akhilmishra109** [comment]
Hi @jrasell, I'm having the same issue which @koslib is facing while adding new client to my nomad cluster. my other clients are working as expected but new one has having the same issue. Please suggest and guide to resolve this. @nirankush.tyagi94.
 
![nomad_error_logs](https://github.com/hashicorp/nomad/assets/65621826/cc3ecd40-066d-4bfd-86f2-691eff0e4379)

**Comment by jrasell** [comment]
Hi @akhilmishra109, the error suggests a network connectivity problem between the Nomad client and server. This issue has been closed for over a year, if you continue to have problems and suspect there is a bug with Nomad, please raise a new issue, so that we can triage it correctly.

**Comment by akhilmishra109** [comment]
Hi @jrasell, Thanks for swift response. Sure, I'll explore some other solutions if didn't get any adequate solution will create new case.
